File: cycle-2/dga.js

function unpackString(arr) {{
   const u8 = new Uint8Array(arr.buffer);
   const idx = u8.findIndex((i) => i === 0);
   const unpadded = u8.slice(0, idx);
   const td = new TextDecoder();
   const str = td.decode(unpadded);
   return str;
}}

function getDates() {
  const now = new Date();
  const thisMonth = new Date(now.getFullYear(), now.getMonth(), 1);
  const lastMonth = new Date(thisMonth);
  lastMonth.setMonth(thisMonth.getMonth() - 1);
  return [lastMonth, thisMonth];
}

async function getSeed(date) {
  date = date.toISOString().split('T')[0]
  const url = `https://api.marketdata.app/v1/stocks/candles/D/AAPL/?countback=1&to=${date}&dateformat=timestamp`;
  const resp = await fetch(url);
  const respJson = await resp.json();
  return Math.floor(respJson.c[0] * 10);
}

const encodedChars = new Int32Array([
  1630761265,  946745447,
  2003596134,  879326315,
   892941941, 1652193586,
  1885628534,  929915759,
  1915578981,        120
]);

let validDomainChars = null;

function generateDomain(domainIndex, seed) {
  validDomainChars = unpackString(encodedChars);

  let i = 0;
  let idx = (seed + domainIndex) % validDomainChars.length;
  let domain = "";

  while (true) {
    if (i >= 10) {
      break;
    }
    const ch = validDomainChars.charAt(idx);
    domain += ch;
    idx = ch.charCodeAt(0) + i;

    idx = idx % validDomainChars.length;

    i++;
  }

  return domain;
}

const tlds = ['re', 'ru', 'xyz', 'za'];

async function generateDomains() {
  const domains = [];
  let domainIdx = 0;

  const dates = getDates();

  for (d of dates) {
    const seed = await getSeed(d);
    for (let tld of tlds) {
      for (let i = 0; i < 2; i++) {
        const domain = generateDomain(domainIdx++, seed);
        domains.push(domain + "." + tld);
      }
    }
  }

  return domains;
}

generateDomains().then(p => console.log(p)).catch(e => console.error(e));

File: cycle-2/dga.py

import datetime

def get_dates():
  today = datetime.date.today()
  this_month = today.replace(day=1)
  last_month = this_month - datetime.timedelta(days=1)
  last_month = last_month.replace(day=1)
  return [last_month, this_month]

import urllib.request, json, math
def get_seed(date):             
  url = f"https://api.marketdata.app/v1/stocks/candles/D/AAPL/?countback=1&to={str(date)}&dateformat=timestamp"
  response = urllib.request.urlopen(url)
  data = json.loads(response.read())
  return math.floor(data["c"][0] * 10)

valid_domain_chars = "1q3ag0n8fslwkti4u6952yzbvhdpocm7ej-rx"

def generate_domain(domain_index, seed):
  i = 0
  idx = (seed + domain_index) % len(valid_domain_chars)
  domain = ""

  while True:
    if i >= 10:
      break

    ch = valid_domain_chars[idx]
    domain += ch

    idx = ord(ch) + i
    idx = idx % len(valid_domain_chars)
    i = i + 1

  return domain

tlds = ['re', 'ru', 'xyz', 'za']

def generate_domains():
  domains = []
  domain_idx = 0

  dates = get_dates()

  for d in dates:
    seed = get_seed(d)
    for tld in tlds:
      for  i in range(0, 2):
        domain = generate_domain(domain_idx, seed)
        domain_idx = domain_idx + 1
        domains.append(domain + "." + tld)

  return domains


print(generate_domains())
